docs/extending.md

```diff
@@ -0,0 +1,299 @@
+```
+
+## Writing Plugins
+
+To write a new Plugin, make a class that inherits from `SnowfakeryPlugin` and implements either the `custom_functions()` method or a `Functions` nested class. The nested class is simple: each method represents a function to expose in the namespace. In this case the function name would be `DoublingPlugin.double`.
+
+```python
+class DoublingPlugin(SnowfakeryPlugin):
+    class Functions:
+        def double(self, value):
+            return value * 2
+```
+
+Alternately, you can implement the custom_functions method to return an
+object with the attributes that implement your namespace:
+
+```python
+class Doubler:
+  def double(self, value):
+      return value * 2
+```
+
+```python
+class DoublingPlugin(SnowfakeryPlugin):
+    def custom_functions(self, *args, **kwargs):
+        return Doubler()
+```
+
+Make sure to accept `*args` and `**kwargs` to allow for future extensibility of the method signature.
+
+Despite the name, plugins can also include data values rather than functions in either form of plugin. Plugins essentially use Python's `getattr` to find attributes, properties, methods or functions in the namespace of the object you return from `custom_functions()`.
+
+Plugin functions can store persistent information in a Python dictionary called self.context.context_vars(). It will always be supplied to your plugin. For example, here is a simple plugin that counts:
+
+```python
+class PluginThatCounts(SnowfakeryPlugin):
+    class Functions:
+        def count(self):
+            context_vars = self.context.context_vars()
+            context_vars.setdefault("count", 0)
+            context_vars["count"] += 1
+            return context_vars["count"]
+```
+
+Plugins also have access to a dictionary called `self.context.field_vars()` which
+represents the values that would be available to a formula running in the same context.
+
+Plugins can return normal Python primitive types, datetime.date, `ObjectRow` or `PluginResult` objects. `ObjectRow` objects represent new output records/objects. `PluginResult` objects
+expose a namespace that other code can access through dot-notation. PluginResults can be
+initialized with either a dict or an object that exposes the namespace through Python
+getattr().
+
+If your plugin generates some special kind of data value which should be serializable
+as a primitive type (usually a string), subclass PluginResult and add a `simplify`
+method to your PluginResult. That method should return a Python primitive value.
+
+In the rare event that a plugin has a function which need its arguments to be passed to it unevaluated, for later (perhaps conditional) evaluation, you can use the `@snowfakery.lazy decorator`. Then you can evaluate the arguments with `self.context.evaluate()`.
+
+For example:
+
+```python
+class DoubleVisionPlugin(SnowfakeryPlugin):
+    class Functions:
+        @lazy
+        def do_it_twice(self, value):
+            "Evaluates its argument 0 times or twice"
```

Comment:
How could it evaluate 0 times?

Reply:
https://github.com/SFDO-Tooling/Snowfakery/pull/353/commits/63e125ab927e8427ae9f5956554f86e8b0e842f8#diff-46aca7079be2ef59a99745c7589744a1171129649b89add013cb0c6e1b2bd8b7R137